pkg/license/license.go

```diff
@@ -0,0 +1,386 @@
+					Name:   info.Name(),
+					Path:   filePath,
+					Weight: weight,
+				})
+				break
+			}
+		}
+
+		return nil
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	// Sort the license files by their 'score' (weight)
+	sort.Slice(licenseFiles, func(i, j int) bool {
+		return licenseFiles[i].Weight > licenseFiles[j].Weight
+	})
+
+	return licenseFiles, nil
+}
+
+// melangeLicenseToLicense is a helper that converts a melange license info to a License struct.
+func melangeLicenseToLicense(license string, licensePath string, override string) License {
+	return License{
+		Name:       license,
+		Type:       "",
+		Confidence: 0.0,
+		Source:     licensePath,
+		Overrides:  override,
+	}
+}
+
+// CollectLicenseInfo collects license information from the given filesystem.
+func CollectLicenseInfo(ctx context.Context, fsys fs.FS) ([]License, error) {
+	log := clog.FromContext(ctx)
+
+	// Find all license-text files
+	licenseFiles, err := FindLicenseFiles(fsys)
+	if err != nil {
+		return nil, fmt.Errorf("finding license files: %w", err)
+	}
+
+	if len(licenseFiles) == 0 {
+		// No license files detected, no linting performed.
+		return nil, nil
+	}
+
+	classifier, err := NewClassifier()
+	if err != nil {
+		return nil, fmt.Errorf("creating classifier: %w", err)
+	}
+
+	melangeClassifier := classifier.(*melangeClassifier)
+	detectedLicenses := []License{}
+	for _, lf := range licenseFiles {
+		dl, err := melangeClassifier.Identify(fsys, lf.Path)
+		if err != nil {
+			return nil, fmt.Errorf("identifying license: %w", err)
+		}
+
+		log.Debugf("detected licenses %v in %s", dl, lf.Path)
+		detectedLicenses = append(detectedLicenses, dl...)
+	}
+
+	return detectedLicenses, nil
+}
+
+func isLicenseMatchConfident(dl License) bool {
+	// This is heuristics, but we want to ignore licenses with a confidence lower than a threshold
+	// We'll make this configurable in the future
+	return dl.Confidence >= 0.9
+}
+
+// LicenseCheck checks the licenses of the files in the given filesystem against the melange configuration.
+func LicenseCheck(ctx context.Context, cfg *config.Configuration, fsys fs.FS) ([]License, []LicenseDiff, error) {
+	log := clog.FromContext(ctx)
+	log.Infof("checking license information")
+
+	detectedLicenses, err := CollectLicenseInfo(ctx, fsys)
+	if err != nil {
+		return nil, nil, fmt.Errorf("collecting license info: %w", err)
+	}
+
+	if detectedLicenses == nil {
+		log.Infof("no license files detected")
+		return nil, nil, nil
+	}
+
+	// Print out all the gathered licenses
+	for _, dl := range detectedLicenses {
+		s := ""
+		// This is heuristics, but we want to ignore licenses with a confidence lower than a threshold
+		if isLicenseMatchConfident(dl) {
+			s = " ignored"
+		}
+		log.Infof("  %s: %s (%f%s) (%s)", dl.Source, dl.Name, dl.Confidence, s, dl.Type)
+	}
+
+	// TODO: Handle low-confidence licenses, possibly by printing out info about those separately!
+
+	diffs := []LicenseDiff{}
+	if cfg != nil {
```

Comment:
The rest of this function might benefit from early returns (or pulling some out into a separate function) to undo some of the deep nesting?